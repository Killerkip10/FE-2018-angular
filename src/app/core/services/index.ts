export * from './auth.service';
export * from './profile.service';
export * from './token.service';
export * from './search-user.service';
