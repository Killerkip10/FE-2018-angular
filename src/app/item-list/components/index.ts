export * from './item/item.component';
export * from './search/search.component';
export * from './path/path.component';
