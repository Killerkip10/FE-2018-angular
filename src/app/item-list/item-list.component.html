<div>
  <app-search (search)="onSearch($event)"></app-search>

  <app-path [path]="path"></app-path>

  <div class="list">
    <app-item
      *ngFor="let value of nodes"
      [node]="value"
      (select)="onSelect($event)">
    </app-item>
  </div>
</div>
