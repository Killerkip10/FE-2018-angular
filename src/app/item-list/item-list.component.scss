.list{
  white-space: nowrap;
  overflow: auto;
}
