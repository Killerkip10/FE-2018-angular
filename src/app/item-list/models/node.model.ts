export interface Node {
  value: string;
  open: boolean;
  children: Node[];
}
