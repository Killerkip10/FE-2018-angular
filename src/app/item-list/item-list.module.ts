import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {FormsModule} from '@angular/forms';

import {ItemListComponent} from './item-list.component';
import {ItemService} from './services';
import {
  ItemComponent,
  SearchComponent,
  PathComponent
} from './components';
import {
  MatInputModule,
  MatFormFieldModule
} from '@angular/material';

@NgModule({
  imports: [
    CommonModule,
    FormsModule,

    MatInputModule,
    MatFormFieldModule
  ],
  declarations: [
    ItemListComponent,
    ItemComponent,
    SearchComponent,
    PathComponent
  ],
  providers: [
    ItemService
  ],
  exports: [
    ItemListComponent
  ]
})
export class ItemListModule { }
