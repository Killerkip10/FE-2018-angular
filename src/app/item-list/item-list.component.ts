import {ChangeDetectionStrategy, Component, OnInit} from '@angular/core';

import {ItemService} from './services';
import {Node} from './models';

@Component({
  selector: 'app-item-list',
  templateUrl: './item-list.component.html',
  styleUrls: ['./item-list.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ItemListComponent implements OnInit {
  public nodes: Node[];
  public path: string;

  constructor(public itemService: ItemService) {}

  ngOnInit(): void {
    this.nodes = this.itemService.getItems();
  }

  public onSelect(node: Node): void {
    node.children = this.itemService.getItems();
    this.path = this.itemService.pathItem(this.nodes, node);
  }
  public onSearch(value: string): void {
    this.nodes = this.itemService.searchItem(this.nodes, value);
  }
}
