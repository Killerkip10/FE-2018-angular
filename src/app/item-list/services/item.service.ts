import {Injectable} from '@angular/core';

import {Node} from '../models';

@Injectable()
export class ItemService {
  constructor() {}

  public getItems(): Node[] {
    return Array.from({length: 100}, (v, i) => ({value: `item ${i}`, open: true, children: null}));
  }
  public searchItem(nodes: Node[], value: string): Node[] {
    this.allOpenItems(nodes);

    if (value) {
      this.searchHelper(nodes, value);
    }

    return nodes;
  }
  public pathItem(nodes: Node[], node: Node): string {
    let path = '';

    for (const key in nodes) {
      if (nodes[key] === node) {
        return node.value;
      }
      if (nodes[key].children) {
        path = this.pathItem(nodes[key].children, node);
      }
      if (path) {
        return `${nodes[key].value} // ${path}`;
      }
    }
  }

  private searchHelper(nodes: Node[], value: string): boolean {
    let isValue = false;

    for (const key in nodes) {
      if (nodes[key].children) {
        if (this.searchHelper(nodes[key].children, value)) {
          isValue = true;
          continue;
        }
      }
      if (nodes[key].value !== value) {
        nodes[key] = Object.assign({}, nodes[key], {open: false});
      } else {
        isValue = true;
      }
    }

    return isValue;
  }
  private allOpenItems(nodes: Node[]): void {
    for (const key in nodes) {
      if (nodes[key].children) {
        this.allOpenItems(nodes[key].children);
      }

      nodes[key] = Object.assign({}, nodes[key], {open: true});
    }
  }
}
