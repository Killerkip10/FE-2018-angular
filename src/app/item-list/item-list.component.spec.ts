import {async, ComponentFixture, TestBed} from '@angular/core/testing';
import {ChangeDetectionStrategy, NO_ERRORS_SCHEMA} from '@angular/core';
import {CommonModule} from '@angular/common';
import {TranslateModule} from '@ngx-translate/core';
import {FormsModule} from '@angular/forms';

import {ItemService} from './services';
import {CoreModule} from '../core/core.module';
import {AppTranslate} from '../app.translate';
import {Node} from './models';
import {ItemListComponent} from './item-list.component';

describe('ItemListComponent', () => {
  let component: ItemListComponent;
  let fixture: ComponentFixture<ItemListComponent>;

  interface ServiceTest {
    getItems: () => Node[];
  }
  class ItemServiceTest implements ServiceTest {
    getItems(): Node[] {
      return [
        {
          value: 'item 1',
          open: true,
          children: null
        },
        {
          value: 'item 2',
          open: true,
          children: null
        }
      ];
    }
  }

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        CoreModule,
        CommonModule,
        FormsModule,
        TranslateModule,

        AppTranslate
      ],
      declarations: [
        ItemListComponent
      ],
      providers: [
        {provide: ItemService, useClass: ItemServiceTest}
      ],
      schemas: [
        NO_ERRORS_SCHEMA
      ]
    })
      .overrideComponent(ItemListComponent, {set: {changeDetection: ChangeDetectionStrategy.Default}})
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(ItemListComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
