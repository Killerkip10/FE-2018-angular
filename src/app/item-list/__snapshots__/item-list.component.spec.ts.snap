// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemListComponent snapshot 1`] = `
<app-item-list
  itemService={[Function ItemServiceTest]}
  nodes={[Function Array]}
>
  <div>
    <app-search />
    <app-path />
    <div
      class="list"
    >
      
      <app-item />
      <app-item />
    </div>
  </div>
</app-item-list>
`;
